src/menu/index.tsx

```diff
@@ -213,23 +183,137 @@ export const Menu = factory(function({
 		);
 	}
 
-	function onActive(index: number, itemDimensions: DimensionResults) {
-		const { position: itemPosition, size: itemSize } = itemDimensions;
-		const { position: rootPosition, size: rootSize } = dimensions.get('root');
-		if (itemPosition.bottom > rootPosition.bottom) {
-			const numInView = Math.ceil(rootSize.height / itemSize.height);
-			icache.set('itemToScroll', Math.max(index - numInView + 1, 0));
-		} else if (itemPosition.top < rootPosition.top) {
-			icache.set('itemToScroll', index);
-		}
+	function renderItem(index: number) {
+		const { value, label, disabled = false } = options[index];
+		const selected = value === selectedValue;
+		const active = index === computedActiveIndex;
+		const itemProps = {
+			widgetId: `${idBase}-item-${index}`,
+			key: `item-${index}`,
+			onSelect: () => {
+				setValue(value);
+			},
+			active,
+			onRequestActive: () => {
+				if (focus.isFocused('root') || !focusable) {
+					setActiveIndex(index);
+				}
+			},
+			disabled
+		};
+
+		const children = itemRenderer
+			? itemRenderer({
+					value,
+					label,
+					disabled,
+					active,
+					selected
+			  })
+			: label || value;
+
+		return listBox ? (
+			<ListBoxItem
+				{...itemProps}
+				selected={selected}
+				theme={theme.compose(
+					listBoxItemCss,
+					css,
+					'item'
+				)}
+			>
+				{children}
+			</ListBoxItem>
+		) : (
+			<MenuItem
+				{...itemProps}
+				theme={theme.compose(
+					menuItemCss,
+					css,
+					'item'
+				)}
+			>
+				{children}
+			</MenuItem>
+		);
 	}
 
-	const itemToScroll = icache.get('itemToScroll');
+	if (initialValue !== undefined && initialValue !== icache.get('initial')) {
+		icache.set('initial', initialValue);
+		icache.set('value', initialValue);
+		icache.set('activeIndex', findIndex(options, (option) => option.value === initialValue));
+	}
+
+	if (itemsInView !== icache.get('itemsInView')) {
+		icache.set('itemsInView', itemsInView);
+
+		const offscreenItemProps = {
+			selected: false,
+			onSelect: () => {},
+			active: false,
+			onRequestActive: () => {},
+			onActive: () => {},
+			scrollIntoView: false,
+			widgetId: 'offcreen'
+		};
+
+		const menuItemChild = itemRenderer
+			? itemRenderer({
+					selected: false,
+					active: false,
+					value: 'offscreen',
+					disabled: false
+			  })
+			: 'offscreen';
+
+		const offscreenMenuItem = <MenuItem {...offscreenItemProps}>{menuItemChild}</MenuItem>;
```

Comment:
What if it's a `ListBoxItem`?

Reply:
ATM they're the same size etc, only difference is the aria / role tags and the ability to select them

Reply:
Aren't they styled by the theme? So in theory could be different sizes if someone themes them like that?

Reply:
fair point, good catch, thanks